dll/win32/shell32/utils.cpp

```diff
@@ -31,6 +31,370 @@ OpenEffectiveToken(
+    _In_ LPCITEMIDLIST pidl,
+    _In_opt_ IBindCtx *pBindCtx,
+    _In_ REFIID riid,
+    _Out_ void **ppvObj)
+{
+    IShellFolder *psfDesktop = NULL;
+
+    *ppvObj = NULL;
+
+    if (!pShellFolder)
+    {
+        SHGetDesktopFolder(&psfDesktop);
+        if (!psfDesktop)
+            return E_FAIL;
+
+        pShellFolder = psfDesktop;
+    }
+
+    HRESULT hr;
+    if (pidl && pidl->mkid.cb)
+        hr = pShellFolder->BindToObject(pidl, pBindCtx, riid, ppvObj);
+    else
+        hr = pShellFolder->QueryInterface(riid, ppvObj);
+
+    if (psfDesktop)
+        psfDesktop->Release();
+
+    if (SUCCEEDED(hr) && !*ppvObj)
+        hr = E_FAIL;
+
+    return hr;
+}
+
+HRESULT SHBindToObject(
+    _In_opt_ IShellFolder *psf,
+    _In_ LPCITEMIDLIST pidl,
+    _In_ REFIID riid,
+    _Out_ void **ppvObj)
+{
+    return SHBindToObjectEx(psf, pidl, NULL, riid, ppvObj);
+}
+
+HRESULT
+Shell_DisplayNameOf(
+    _In_ IShellFolder *psf,
+    _In_ LPCITEMIDLIST pidl,
+    _In_ DWORD dwFlags,
+    _Out_ LPWSTR pszBuf,
+    _In_ UINT cchBuf)
+{
+    *pszBuf = UNICODE_NULL;
+    STRRET sr;
+    HRESULT hr = psf->GetDisplayNameOf(pidl, dwFlags, &sr);
+    if (FAILED(hr))
+        return hr;
+    return StrRetToBufW(&sr, pidl, pszBuf, cchBuf);
+}
+
+DWORD
+SHGetAttributes(_In_ IShellFolder *psf, _In_ LPCITEMIDLIST pidl, _In_ DWORD dwAttributes)
+{
+    LPCITEMIDLIST pidlLast;
+
+    if (psf)
+    {
+        psf->AddRef();
+        pidlLast = pidl;
+    }
+    else
+    {
+        SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &psf), &pidlLast);
+    }
+
+    if (!psf)
+        return 0;
+
+    DWORD oldAttrs = dwAttributes;
+    if (FAILED(psf->GetAttributesOf(1, &pidlLast, &dwAttributes)))
+        dwAttributes = 0;
+    else
+        dwAttributes &= oldAttrs;
+
+    if ((dwAttributes & SFGAO_FOLDER) &&
+        (dwAttributes & SFGAO_STREAM) &&
+        !(dwAttributes & SFGAO_STORAGEANCESTOR) &&
+        (oldAttrs & SFGAO_STORAGEANCESTOR) &&
+        (SHGetObjectCompatFlags(psf, NULL) & 0x200))
+    {
+        dwAttributes &= ~(SFGAO_STREAM | SFGAO_STORAGEANCESTOR);
+        dwAttributes |= SFGAO_STORAGEANCESTOR;
+    }
+
+    if (psf)
+        psf->Release();
+
+    return dwAttributes;
+}
+
+HRESULT SHCoInitialize(VOID)
+{
+    HRESULT hr = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+    if (FAILED(hr))
+        hr = CoInitializeEx(NULL, COINIT_DISABLE_OLE1DDE);
+    return hr;
+}
+
+HRESULT
+SHBindToFolderIDListParent(
+    _In_opt_ IShellFolder *psfRoot,
+    _In_ LPCITEMIDLIST pidl,
+    _In_ REFIID riid,
+    _Out_ void **ppv,
+    _Out_opt_ LPCITEMIDLIST *ppidlLast)
+{
+    HRESULT hr = E_OUTOFMEMORY;
+
+    LPITEMIDLIST pidlClone = ILCloneParent(pidl);
+    if (pidlClone)
+    {
+        hr = SHBindToObjectEx(psfRoot, pidlClone, NULL, riid, ppv);
+        ILFree(pidlClone);
+    }
+
+    if (ppidlLast)
+        *ppidlLast = ILFindLastID(pidl);
+
+    return hr;
+}
+
+HRESULT
+SHBindToIDListParent(
```

Comment:
How is this different from SHBindToParent that we already have?

Reply:
`SHBindToIDListParent` and `SHBindToParent` looks nearly identical.

Reply:
`_ILIsPidlSimple(pidl)` and `!(pidl && pidl->mkid.cb)` are different conditions.

Reply:
Yes, `SHBindToIDListParent` and `SHBindToParent` is identical.
`ILCloneParent` will absorb the difference between `_ILIsPidlSimple(pidl)` and `_ILIsDesktop(pidl)`.

Reply:
SHBindToFolderIDListParent clones the parent so the pidlsimple case becomes isdesktop there.

Reply:
Deleted `SHBindToFolderIDListParent` and `SHBindToIDListParent` functions.

Reply:
SHBindToFolderIDListParent and SHBindToFolderIDListParentEx are documented, the question becomes, does SHBindToParent simply call it?

Reply:
I think `SHBindToParent` and `SHBindToIDListParent` are the identically-implemented function.
`SHBindToFolderIDListParent` call will be `SHBindToParent` function call.